### ISSUE TITLE
400 Bad Request: token expired

### ISSUE THREAD

**Issue by dlaprade** [craftcms/aws-s3#22]
I wrote a plugin that creates Entries that contain image assets. After around 1k entires, it dies with error:

```
Exception 'Aws\S3\Exception\S3Exception' with message 'Error executing "ListObjects" on "***"; AWS HTTP error: Client error: 'GET ***' resulted in a '400 Bad Request' response:
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>ExpiredToken</Code><Message>The provided token has expired.</Message (truncated...)
 ExpiredToken (client): The provided token has expired. - <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>ExpiredToken</Code><Message>The provided token has expired.</Message>
```

I added a try/catch in order to retry the save after a sleep, but it was not helpful.  Is there a better way to handle this or is this a bug?

INFO:
CraftCMS: 3.0
PHP: 7.0

**Comment by andris-sevcenko** [comment]
Are you just creating one instance of the Volume and just re-using it for all the entries? It's very well possible that the AWS STS token expires during that.

How long does it take for the plugin to reach 1K entries?

**Comment by dlaprade** [comment]
I created a separate function that handles the upload of an image:

```
private function _uploadNewImageAsset( string $filename, string $tempLocation, $folderId, bool $makeUnique ): Asset
    {
$assets = Craft::$app->getAssets();
$folder = $assets->findFolder( [ 'id' => $folderId ] );
...
$asset = new Asset();
$asset->tempFilePath = $tempLocation . '/' . $filename;
$asset->filename         = $filename;
$asset->newFolderId  = $folder->id;
$asset->volumeId        = $folder->volumeId;
$asset->kind                = Asset::KIND_IMAGE;
$asset->setScenario( Asset::SCENARIO_CREATE );
...
$result = Craft::$app->getElements()->saveElement( $asset );
...

return $asset
}
```

Is there an issue with doing it this way?

**Comment by bleepsandblops** [comment]
Having the same issue doing a big import using feedme, the token eventually times out...

**Comment by Mosnar** [comment]
I've run into this myself as well. I've tried increasing the default session token lifespan from within AWS; however, that doesn't seem to do the trick, though I may have missed something since I'm using IAM EC2 policies to infer tokens. Given that Craft is requesting a 60 minute token and caching it for that long but it seems to expire around the 15 minute mark (the minimum lifespan of an STS token), it seems likely that AWS is giving us a token shorter lived than what we're requesting/expecting. A solution might be to utilize the returned token expiration date to derive a dynamic cache duration period.

**Comment by khalwat** [comment]
I'm running into this as well with ImageOptimize being run via the CLI. It ultimately uses Craft CMS transforms under the hood, so it goes through the whole Asset Transforms -> AWS S3 plugin pipeline.

@andris-sevcenko is there a way to programatically renew the token if I know I'm working on an AWS S3 volume?

https://github.com/nystudio107/craft-imageoptimize/pull/150

**Comment by khalwat** [comment]
@Mosnar I ended up at the same place you did -- it looks like AWS can apparently return a value different than we ask for, even though what we're asking for falls within the specification [per the documentation](https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-sts-2011-06-15.html#getsessiontoken)

Created a PR https://github.com/craftcms/aws-s3/pull/61

**Comment by andris-sevcenko** [comment]
Merged #61 

I'll cut a release with this in a few hours.
